File: scripts/test-docs-rs-metadata.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'fileutils'
require 'json'
require 'open3'
require 'rubygems/package'
require 'tmpdir'
require 'zlib'

REPOSITORY_ROOT = File.expand_path('..', __dir__)

class ContractError < StandardError; end

def fail_contract(message)
  raise ContractError, "docs.rs metadata contract: #{message}"
end

def cargo_metadata(manifest_path)
  output, status = Open3.capture2e(
    'cargo', 'metadata', '--locked', '--no-deps', '--format-version=1', '--manifest-path', manifest_path,
    chdir: REPOSITORY_ROOT
  )
  fail_contract("cargo metadata failed:\n#{output}") unless status.success?

  JSON.parse(output)
end

def package_for_manifest(metadata, manifest)
  canonical_manifest = File.realpath(manifest)
  package = metadata.fetch('packages').find do |candidate|
    File.realpath(candidate.fetch('manifest_path')) == canonical_manifest
  end
  fail_contract("cargo metadata did not return #{canonical_manifest}") unless package

  package
end

def docs_rs_features(manifest, document: File.read(manifest))
  header = '[package.metadata.docs.rs]'
  start = document.index("#{header}\n")
  fail_contract("#{manifest} must contain #{header}") unless start

  table = document[(start + header.length + 1)..]
  table = table.split(/^\[/, 2).fetch(0)
  match = table.match(/^features\s*=\s*\[(?<features>[^\]]*)\]\s*$/)
  fail_contract("#{manifest} docs.rs metadata must define features") unless match

  match[:features].scan(/"([^"]+)"/).flatten
end

def assert_docs_rs_features(metadata, manifest, document: File.read(manifest))
  features = docs_rs_features(manifest, document: document)
  expected = ['mdx']
  fail_contract("#{manifest} docs.rs features must be #{expected.inspect}, got #{features.inspect}") unless features == expected

  package_features = package_for_manifest(metadata, manifest).fetch('features')
  fail_contract("#{manifest} must declare the mdx feature") unless package_features.key?('mdx')
end

def package_crate(target_directory)
  output, status = Open3.capture2e(
    'cargo', 'package', '--allow-dirty', '--locked', '--offline', '--no-verify', '--target-dir', target_directory,
    chdir: REPOSITORY_ROOT
  )
  fail_contract("cargo package failed:\n#{output}") unless status.success?

  manifest = File.join(REPOSITORY_ROOT, 'Cargo.toml')
  package = package_for_manifest(cargo_metadata(manifest), manifest)
  crate = File.join(target_directory, 'package', "#{package.fetch('name')}-#{package.fetch('version')}.crate")
  fail_contract("cargo package did not create #{crate}") unless File.file?(crate)

  crate
end

def build_mdx_docs(target_directory)
  output, status = Open3.capture2e(
    { 'RUSTDOCFLAGS' => '-D warnings' },
    'cargo', 'doc', '--locked', '--no-deps', '--features', 'mdx', '--target-dir', target_directory,
    chdir: REPOSITORY_ROOT
  )
  fail_contract("MDX documentation build failed:\n#{output}") unless status.success?

  page = File.join(target_directory, 'doc', 'ferromark', 'mdx', 'index.html')
  fail_contract("MDX documentation build did not generate #{page}") unless File.file?(page)
end

def extract_crate(crate, destination)
  Zlib::GzipReader.open(crate) do |gzip|
    Gem::Package::TarReader.new(gzip) do |tar|
      tar.each do |entry|
        next unless entry.file?

        path = File.join(destination, entry.full_name)
        FileUtils.mkdir_p(File.dirname(path))
        File.binwrite(path, entry.read)
      end
    end
  end
end

def validate(repository_root)
  source_manifest = File.join(repository_root, 'Cargo.toml')
  assert_docs_rs_features(cargo_metadata(source_manifest), source_manifest)

  Dir.mktmpdir('ferromark-docs-rs-docs-target.') do |target_directory|
    build_mdx_docs(target_directory)
  end

  Dir.mktmpdir('ferromark-docs-rs-package.') do |destination|
    Dir.mktmpdir('ferromark-docs-rs-target.') do |target_directory|
      extract_crate(package_crate(target_directory), destination)
      manifests = Dir.glob(File.join(destination, '*', 'Cargo.toml'))
      fail_contract("expected one packaged Cargo.toml, got #{manifests.inspect}") unless manifests.length == 1
      packaged_manifest = manifests.fetch(0)
      assert_docs_rs_features(cargo_metadata(packaged_manifest), packaged_manifest)
    end
  end
end

def assert_rejected(label)
  yield
  raise "#{label} mutation unexpectedly passed"
rescue ContractError
  # Expected: this mutation must be rejected by the contract.
end

def self_test(repository_root)
  source_manifest = File.join(repository_root, 'Cargo.toml')
  source_document = File.read(source_manifest)
  metadata = cargo_metadata(source_manifest)
  metadata_without_mdx = JSON.parse(JSON.generate(metadata))
  package_for_manifest(metadata_without_mdx, source_manifest).fetch('features').delete('mdx')

  mutations = {
    'missing docs.rs table' => source_document.sub("[package.metadata.docs.rs]\nfeatures = [\"mdx\"]\n\n", ''),
    'broader docs.rs feature set' => source_document.sub('features = ["mdx"]', 'features = ["mdx", "profiling"]')
  }
  mutations.each do |label, document|
    assert_rejected(label) { assert_docs_rs_features(metadata, source_manifest, document: document) }
  end
  assert_rejected('undeclared mdx feature') { assert_docs_rs_features(metadata_without_mdx, source_manifest) }
end

begin
  if ARGV == ['--self-test']
    validate(REPOSITORY_ROOT)
    self_test(REPOSITORY_ROOT)
  elsif ARGV.empty?
    validate(REPOSITORY_ROOT)
  else
    abort 'usage: test-docs-rs-metadata.rb [--self-test]'
  end
rescue ContractError => error
  abort error.message
end

puts 'docs.rs metadata and packaged manifest checks passed'

File: scripts/test-docs-rs-metadata.mjs
import assert from "node:assert/strict";
import { spawnSync } from "node:child_process";
import { mkdtempSync, readFileSync, realpathSync, rmSync, statSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { describe, it } from "node:test";

import { ContractError, repositoryPath, repositoryRoot } from "./lib/contracts.mjs";

function failContract(message) {
  throw new ContractError(`docs.rs metadata contract: ${message}`);
}

function run(command, args, options = {}) {
  const result = spawnSync(command, args, {
    cwd: repositoryRoot,
    encoding: "utf8",
    ...options,
    env: { ...process.env, ...options.env },
  });
  if (result.error) {
    failContract(`${command} could not be started: ${result.error.message}`);
  }
  return result;
}

function cargoMetadata(manifestPath) {
  const result = run("cargo", [
    "metadata",
    "--locked",
    "--no-deps",
    "--format-version=1",
    "--manifest-path",
    manifestPath,
  ]);
  if (result.status !== 0) {
    failContract(`cargo metadata failed:\n${result.stdout}${result.stderr}`);
  }
  return JSON.parse(result.stdout);
}

function packageForManifest(metadata, manifest) {
  const canonicalManifest = realpathSync(manifest);
  const found = metadata.packages.find(
    (candidate) => realpathSync(candidate.manifest_path) === canonicalManifest,
  );
  if (!found) {
    failContract(`cargo metadata did not return ${canonicalManifest}`);
  }
  return found;
}

function docsRsTable(manifest, document) {
  const header = "[package.metadata.docs.rs]";
  const start = document.indexOf(`${header}\n`);
  if (start === -1) {
    failContract(`${manifest} must contain ${header}`);
  }
  return document.slice(start + header.length + 1).split(/^\[/m)[0];
}

function docsRsFeatures(manifest, document) {
  const match = docsRsTable(manifest, document).match(/^features\s*=\s*\[([^\]]*)\]\s*$/m);
  if (!match) {
    failContract(`${manifest} docs.rs metadata must define features`);
  }
  return [...match[1].matchAll(/"([^"]+)"/g)].map((entry) => entry[1]);
}

function docsRsNoDefaultFeatures(manifest, document) {
  const match = docsRsTable(manifest, document).match(
    /^no-default-features\s*=\s*(true|false)\s*$/m,
  );
  if (!match) {
    failContract(`${manifest} docs.rs metadata must define no-default-features`);
  }
  return match[1] === "true";
}

function assertDocsRsFeatures(metadata, manifest, document = readFileSync(manifest, "utf8")) {
  const features = docsRsFeatures(manifest, document);
  const expected = ["mdx"];
  if (
    features.length !== expected.length ||
    features.some((name, index) => name !== expected[index])
  ) {
    failContract(
      `${manifest} docs.rs features must be ${JSON.stringify(expected)}, got ${JSON.stringify(features)}`,
    );
  }
  if (!docsRsNoDefaultFeatures(manifest, document)) {
    failContract(`${manifest} docs.rs must disable default features`);
  }
  if (!("mdx" in packageForManifest(metadata, manifest).features)) {
    failContract(`${manifest} must declare the mdx feature`);
  }
}

function packageCrate(targetDirectory) {
  const result = run("cargo", [
    "package",
    "--allow-dirty",
    "--locked",
    "--offline",
    "--no-verify",
    "--target-dir",
    targetDirectory,
  ]);
  if (result.status !== 0) {
    failContract(`cargo package failed:\n${result.stdout}${result.stderr}`);
  }
  const manifest = repositoryPath("Cargo.toml");
  const packageInfo = packageForManifest(cargoMetadata(manifest), manifest);
  const crate = path.join(
    targetDirectory,
    "package",
    `${packageInfo.name}-${packageInfo.version}.crate`,
  );
  if (!statSync(crate, { throwIfNoEntry: false })?.isFile()) {
    failContract(`cargo package did not create ${crate}`);
  }
  return crate;
}

function buildMdxDocs(targetDirectory) {
  const result = run(
    "cargo",
    [
      "doc",
      "--locked",
      "--no-deps",
      "--no-default-features",
      "--features",
      "mdx",
      "--target-dir",
      targetDirectory,
    ],
    { env: { RUSTDOCFLAGS: "-D warnings" } },
  );
  if (result.status !== 0) {
    failContract(`MDX documentation build failed:\n${result.stdout}${result.stderr}`);
  }
  const page = path.join(targetDirectory, "doc", "ferromark", "mdx", "index.html");
  if (!statSync(page, { throwIfNoEntry: false })?.isFile()) {
    failContract(`MDX documentation build did not generate ${page}`);
  }
}

function extractCrate(crate, destination) {
  const result = spawnSync("tar", ["-xzf", crate, "-C", destination], { encoding: "utf8" });
  if (result.error || result.status !== 0) {
    failContract(`could not extract ${crate}: ${result.error?.message ?? result.stderr}`);
  }
}

function withTemporaryDirectory(prefix, callback) {
  const directory = mkdtempSync(path.join(tmpdir(), prefix));
  try {
    return callback(directory);
  } finally {
    rmSync(directory, { force: true, recursive: true });
  }
}

const sourceManifest = repositoryPath("Cargo.toml");
const sourceDocument = readFileSync(sourceManifest, "utf8");
const metadata = cargoMetadata(sourceManifest);

describe("docs.rs metadata contract", () => {
  it("publishes the mdx feature on docs.rs", () => {
    assertDocsRsFeatures(metadata, sourceManifest);
  });

  it("builds the mdx documentation without rustdoc warnings", () => {
    withTemporaryDirectory("ferromark-docs-rs-docs-target.", buildMdxDocs);
  });

  it("keeps the docs.rs metadata in the packaged manifest", () => {
    withTemporaryDirectory("ferromark-docs-rs-package.", (destination) => {
      withTemporaryDirectory("ferromark-docs-rs-target.", (targetDirectory) => {
        extractCrate(packageCrate(targetDirectory), destination);
        const packagedManifest = path.join(
          destination,
          `ferromark-${packageForManifest(metadata, sourceManifest).version}`,
          "Cargo.toml",
        );
        assert.ok(
          statSync(packagedManifest, { throwIfNoEntry: false })?.isFile(),
          `expected a packaged Cargo.toml at ${packagedManifest}`,
        );
        assertDocsRsFeatures(cargoMetadata(packagedManifest), packagedManifest);
      });
    });
  });

  it("rejects a missing docs.rs table", () => {
    assert.throws(
      () =>
        assertDocsRsFeatures(
          metadata,
          sourceManifest,
          sourceDocument.replace(
            '[package.metadata.docs.rs]\nfeatures = ["mdx"]\nno-default-features = true\n\n',
            "",
          ),
        ),
      ContractError,
    );
  });

  it("rejects a broader docs.rs feature set", () => {
    assert.throws(
      () =>
        assertDocsRsFeatures(
          metadata,
          sourceManifest,
          sourceDocument.replace('features = ["mdx"]', 'features = ["mdx", "profiling"]'),
        ),
      ContractError,
    );
  });

  it("rejects default features on docs.rs", () => {
    assert.throws(
      () =>
        assertDocsRsFeatures(
          metadata,
          sourceManifest,
          sourceDocument.replace("no-default-features = true", "no-default-features = false"),
        ),
      ContractError,
    );
  });

  it("rejects an undeclared mdx feature", () => {
    const withoutMdx = structuredClone(metadata);
    delete packageForManifest(withoutMdx, sourceManifest).features.mdx;
    assert.throws(() => assertDocsRsFeatures(withoutMdx, sourceManifest), ContractError);
  });
});
